synapse/push/push_tools.py

```diff
@@ -49,7 +50,33 @@ async def get_badge_count(store: DataStore, user_id: str, group_by_room: bool) -
 async def get_context_for_event(
     storage: StorageControllers, ev: EventBase, user_id: str
 ) -> Dict[str, str]:
-    ctx = {}
+    ctx: Dict[str, str] = {}
+
+    if ev.internal_metadata.outlier:
+        # We don't have state for outliers, so we can't compute the context
+        # except for invites with `invite_room_state` set..
+        if (
+            ev.type != EventTypes.Member
```

Comment:
I think we could fill in the `sender_display_name` in more cases. Like if the `sender` matches the `state_key`

Reply:
Hmm, I guess knocking is also a thing here. I've expanded the logic a bit.